insights/combiners/rhel_for_edge.py

```diff
@@ -0,0 +1,70 @@
+"""
+Combiner for edge computing systems
+===================================
+This combiner uses the following parsers to determine if the system is an edge computing systems.
+
+* :py:class:`insights.parsers.installed_rpms.InstalledRpms`
+* :py:class:`insights.parsers.cmdline.CmdLine`
+* :py:class:`insights.parsers.systemd.unitfiles.ListUnits`
+* :py:class:`insights.parsers.redhat_release.RedhatRelease`
+"""
+from insights.core.plugins import combiner
+from insights.parsers.installed_rpms import InstalledRpms
+from insights.parsers.cmdline import CmdLine
+from insights.parsers.systemd.unitfiles import ListUnits
+from insights.parsers.redhat_release import RedhatRelease
+
+
+@combiner(InstalledRpms, CmdLine, ListUnits, RedhatRelease)
+class RhelForEdge(list):
+    """
+    Combiner for checking if the system is an edge computing systems.
+    Edge computing system packages are managed via rpm-ostree.
+    Red Hat CoreOS is also managed via rpm-ostree, use the string "Red Hat Enterprise Linux release"
+    from "/etc/redhat-release" to determine if it is an edge computing system as it is "Red Hat Enterprise Linux CoreOS release"
+    on RedHat CoreOs.
+
+    When an edge computing system create by console edge image is configured to use automated management,
+    the output of "rhc status" is the following:
+
+            Connection status for test.localhost:
+            - Connected to Red Hat Subscription Management
+            - The Red Hat connector daemon is active
+
+    If a system can upload insights archive, it must be connected to Red Hat Subscription Management, rhcd service running
+    means an edge computing system is configured to use automated management. Note: it is able to run rhcd service on the
+    edge system created corkpit edge image, "is_automated" is only for front-end resolution surface, it is used when
+    customers determine that the image is from console.
+
+    Examples:
+        >>> type(rhel_for_edge_obj)
+        <class 'insights.combiners.rhel_for_edge.RhelForEdge'>
+        >>> rhel_for_edge_obj.is_edge
+        True
+        >>> rhel_for_edge_obj.is_automated
+        True
+
+    Attributes:
+        is_edge (bool): True when it is an edge computing system
+        is_automated (bool): True when the the edge computing system is configured to use automated management
+    """
+
+    def __init__(self, rpms, cmdline, units, redhatrelease):
+        self.is_edge = False
+        self.is_automated = False
+
+        if 'rpm-ostree' in rpms and 'yum' not in rpms and 'ostree' in cmdline and "red hat enterprise linux release" in redhatrelease.raw.lower():
```

Comment:
Per the current implementation, this combiner will always get triggered and return something. Whether it's Okay to just skip this combiner by raising a "SkipComponent` when a system is not an `edge computing system`? which could save a bit of memory on the server.

Meanwhile, take this combiner an optional `condition` in the rule, only when it's got fire, it's an `edge computing system. What do you think? @wushiqinlou

Reply:
I would prefer to not raise `SkipComponent`. I think that should be raised only if invalid data are provided to the `combiner`? The docstring says "This class should be raised by components that want to be taken out of dependency resolution." If the combiner gets valid data and can make decision if system is or is not Edge system, I do not see a reason why it should remove itself from dependency resolution instead of providing expected answer?

Reply:
It also makes sense to me that skips the combiner only when cannot make the decision due to the lack of required parsers.
I just thought from the Advisor rules perspective, the results are the same, and "nothing" will be done for the advisor rule (when the rule also works for `edge`): no matter when either it's not an `edge` or cannot identify if it's an `edge`.

I have no problem making the combiner always get triggered if it's hard for the ProdSec rules to work in the same way as the Advisor rules.

Reply:
You are right that rules in this case do not care about distinguishing between "not edge" and "do not know". The reason why I do not like it is that when debugging rules you may have hard time finding out if there is a bug in combiner or if really so many systems are "not edge". 

Rules could work with `SkipComponent` by basically writing a workaround in condition (`if combiner is None: this actually means False and you better hope it is not None all the time due to bug`), I think that approach is unexpected from the rule writer perspective and semantically incorrect.

Prodsec rules can handle it either way.

Reply:
I agree with you.